docs/internals/optimizer.rst

```diff
@@ -1096,6 +1097,46 @@ The step LiteralRematerialiser is not required for correctness. It helps deal wi
 ``function f(x) -> y { revert(y, y} }`` where the literal ``y`` will be replaced by its value ``0``,
 allowing us to rewrite the function.
 
+.. _unused-store-eliminator:
+
+UnusedStoreEliminator
+^^^^^^^^^^^^^^^^^^^^^
+
+Optimizer component that removes ``sstore`` and ``mstore`` statements if conditions are met for said removal.
+In case of ``sstore``, if all code paths lead to a revert, or if a previous ``sstore`` is overwritten,
+the statement will be removed. For example, the following code
+
+.. code-block:: yul
+
+    {
+        let c := calldataload(0)
+        sstore(c, 1)
+        if c {
+            sstore(c, 2)
+        }
+        sstore(c, 3)
+    }
+
+will produce the code below after the unused store eliminator step is run
+
+.. code-block:: yul
+
+    {
+        {
+            let c := calldataload(0)
+            let _2 := 1
+            if c { let _3 := 2 }
+            sstore(c, 3)
+        }
+    }
+
+For ``mstore``, things are generally simpler, as all ``mstore`` statements will be removed if they are never
+read from in all code paths.
```

Comment:
Yeah, but that's only half of the truth... that's true for ``mstore``s in the main outermost yul block - if you're analyzing a function body on the other hand, it works more like the storage case: then you don't know if the memory may be read, *after* returning from the function, so anything that's not overwritten in the function itself, will be retained.

Also it's not just ``mstore``s, but any memory-writing operation like ``calldatacopy`` as well for example (in which case you have to deal with arbitrary ranges in memory, for which "overwriting" or "not reading from" is actually rather hard to determine in general, but we don't have to get into that part in the docs anyways).